### ISSUE TITLE
[mcp23017] Binding is not working on OH2.5

### ISSUE THREAD

**Issue by zolee28** [openhab/openhab-addons#6111]
<a alt="bounty" href="https://www.bountysource.com/issues/80686265"><img src="https://api.bountysource.com/badge/issue?issue_id=80686265"></a>

After installing the addon through PaperUI, the following error appears. If someone cold resolve this, it would be nice. The addon right now does not work.

```
2019-09-18 22:28:06.573 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.dispose()' on 'org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler@bdf7ba': null
java.lang.NullPointerException: null
	at org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler.dispose(Mcp23017Handler.java:186) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [133:org.openhab.core:2.5.0.M3]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53) [133:org.openhab.core:2.5.0.M3]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2019-09-18 22:28:06.579 [ERROR] [core.thing.internal.ThingManagerImpl] - Exception occurred while disposing handler of thing 'mcp23017:mcp23017:27599645': null
java.lang.NullPointerException: null
	at org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler.dispose(Mcp23017Handler.java:186) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [133:org.openhab.core:2.5.0.M3]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53) [133:org.openhab.core:2.5.0.M3]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?] 
```

* Version used: Openhab 2.5-M3 installed by Openhabian
* Operating System RPi 3

**Comment by zolee28** [comment]
Bumping the thread with a small bounty.

[https://www.bountysource.com/issues/80686265-mcp23017-binding-is-not-working-on-oh2-5](url)

Posted the issue also in the community forum.

**Comment by 9037568** [comment]
You really should post a full debug log that starts from before the installation of the binding.

I'm guessing the null pointer exception you've captured in the ERROR message is being caused by a configuration problem, which can't be determined based on the information you've provided.

But I'll have a fix uploaded for the null pointer exception shortly.

**Comment by zolee28** [comment]
If you could tell me how to do it, I will.

One hint, the addon version 2.4 works flawlessly and when I switch to 2.5.0-1 the error appears. I managed to get it working somehow by using the ver. 2.4.0 addon putting the jar file here \lib\openhab2\tmp\mvn\org\openhab\binding\org.openhab.binding.mcp23017\ or here usr\share\openhab2\addons - I do not know which one does the trick.

**Comment by 9037568** [comment]
Fixed by #6618.

**Comment by zolee28** [comment]
So I took the Snapshot from here (https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/addons/bundles/org.openhab.binding.mcp23017/2.5.2-SNAPSHOT/), placed in the folder I mentioned above and it's not working.

``

2020-01-29 17:47:08.115 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.handleCommand()' on 'org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler@1aa8eff': null

java.lang.NullPointerException: null

	at org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler.handleOutputCommand(Mcp23017Handler.java:116) ~[?:?]

	at org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler.handleCommand(Mcp23017Handler.java:83) ~[?:?]

	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_222]

	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_222]

	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_222]

	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_222]

	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]

	at org.eclipse.smarthome.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [bundleFile:?]

	at com.sun.proxy.$Proxy190.handleCommand(Unknown Source) [?:?]

	at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:74) [bundleFile:?]

	at org.eclipse.smarthome.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:48) [bundleFile:?]

	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_222]

	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_222]

	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_222]

	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_222]

	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]

	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]

	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_222]

	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_222]

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_222]

	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]
``

**Comment by zolee28** [comment]
So as I mentioned, it works fine with this: org.openhab.binding.mcp23017-2.4.0.jar, can't someone just rename it to 2.5.2 and it's done...

**Comment by 9037568** [comment]
This is a different error than the one you were getting before, @zolee28 .
If I get the time, I might try to look into any changes that might have been made between 2.4.0 and 2.5.x.

If you provide your configuration information and a debug log, we'll reopen this issue for further work.

**Comment by zeogeo** [comment]
Hi,
I updated from OH 2.4 to OH 2.5 and I have the problem in calling handleCommand() method since then:
`2020-02-09 08:07:39.485 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.handleCommand()' on 'org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler@cee989': null
java.lang.NullPointerException: null
	at org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler.handleOutputCommand(Mcp23017Handler.java:116) ~[?:?]
	at org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler.handleCommand(Mcp23017Handler.java:83) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_152]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_152]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_152]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_152]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]
	at org.eclipse.smarthome.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [bundleFile:?]
	at com.sun.proxy.$Proxy446.handleCommand(Unknown Source) [?:?]
	at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:74) [bundleFile:?]
	at org.eclipse.smarthome.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:48) [bundleFile:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_152]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_152]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_152]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_152]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_152]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_152]`

What further information/data would you need to re-open the ticket? I believe it is a generic error, irrelevant of user data and comparing the 2.4.0 vs. 2.5.x would be the way to go. Thank you.

**Comment by 9037568** [comment]
You need to provide your configuration information related to the binding (including items, things, sitemaps) as well as a debug log.

FWIW, I did do a brief review of changes from 2.4 to 2.5, but nothing stood out as a likely problem.

**Comment by zeogeo** [comment]
from configuration files: 
relevant items: 
```
Switch          F2_MasterBedroom_Window_Down "Window Down"      <rollershutter>   (gShutter)      {channel="mcp23017:mcp23017:chip20:output#B4"}
Switch          F2_MasterBedroom_Window_Up  "Window Up"        <rollershutter>   (gShutter)      {channel="mcp23017:mcp23017:chip20:output#B6"}
```

relevant things:
```
Thing mcp23017:mcp23017:chip20  "mcp1" [address=20,bus=1] {
    Type output_pin : output#A0 [default_state="HIGH"]
    Type output_pin : output#A1 [default_state="HIGH"]
    
    Type output_pin : output#B0 [default_state="LOW"]

    Type output_pin : output#B3 [default_state="LOW"]
    Type output_pin : output#B4 [default_state="LOW"]
    Type output_pin : output#B5 [default_state="LOW"]
    Type output_pin : output#B6 [default_state="LOW"]
}
```

rule sending command: 
```
rule "Bedroom Window Down"
when
  Item F2_MasterBedroom_Window received command DOWN
then
  logInfo("blind_bedroom.rules", "Bedroom window Blind down")
  F2_MasterBedroom_Window_Up.sendCommand(OFF) // motor protection
  F2_MasterBedroom_Window_Down.sendCommand(ON)
  var howLongToRun = 20;
  if (winterMode.state == OFF) {
    howLongToRun = 15;
  }
  createTimer(now.plusSeconds(howLongToRun), [ | F2_MasterBedroom_Window_Down.sendCommand(OFF) ])
end
```

ouput of log:tail from karaf, after changing to DEBUG for org.openhab:

```
19:10:39.800 [DEBUG] [3017.internal.handler.Mcp23017Handler] - Received command: ON on channelGroup output on channel B4
19:10:39.810 [DEBUG] [3017.internal.handler.Mcp23017Handler] - Defined channel in group: output. Valid: true
19:10:39.821 [DEBUG] [3017.internal.handler.Mcp23017Handler] - Is channel B4 in supported channels: true
19:10:39.832 [DEBUG] [ternal.handler.Mcp23017PinStateHolder] - Getting output pin for channel mcp23017:mcp23017:chip20:output#B4
19:10:39.842 [DEBUG] [ternal.handler.Mcp23017PinStateHolder] - initializeOutputPin for channel mcp23017:mcp23017:chip20:output#B4
19:10:39.852 [DEBUG] [ternal.handler.Mcp23017PinStateHolder] - initializing for pinState LOW
19:10:39.869 [DEBUG] [ternal.handler.Mcp23017PinStateHolder] - Bound digital output for PIN: GPIO B4, channel: mcp23017:mcp23017:chip20:output#B4, pinState: LOW
19:10:39.881 [ERROR] [rnal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.handleCommand()' on 'org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler@5ff410': null
java.lang.NullPointerException: null
        at org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler.handleOutputCommand(Mcp23017Handler.java:116) ~[?:?]
        at org.openhab.binding.mcp23017.internal.handler.Mcp23017Handler.handleCommand(Mcp23017Handler.java:83) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_152]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_152]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_152]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_152]
        at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]
        at org.eclipse.smarthome.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [bundleFile:?]
        at com.sun.proxy.$Proxy187.handleCommand(Unknown Source) [?:?]
        at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:74) [bundleFile:?]
        at org.eclipse.smarthome.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:48) [bundleFile:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_152]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_152]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_152]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_152]
        at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]
        at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_152]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_152]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_152]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_152]
19:10:40.544 [DEBUG] [hab.binding.http.internal.HttpBinding] - 'execute()' called ...

```

In the meanwhile I also looked into the code and if the problem happens in line 116 (providing the line number didn't change from 2.5.1 to 2.5.2-SNAPSHOT (actual code I forked)), then this is the line:
`boolean activeLowFlag = ACTIVE_LOW_ENABLED.equalsIgnoreCase(configuration.get(ACTIVE_LOW).toString());`

I looked into my configuration (REST API) and I don't see 'active_low' in my channels. 
So either the _configuration_ is null or the _configuration.get(ACTIVE_LOW)_ delivers null, am I right?

**Comment by 9037568** [comment]
>So either the configuration is null or the configuration.get(ACTIVE_LOW) delivers null, am I right?

Yes, because your configuration does not define the `active_low` property:

````
Type output_pin : output#B4 [default_state="LOW"]
````

That also provides a hint as to a workaround for the problem, which is to define the property:

````
Type output_pin : output#B4 [default_state="LOW", active_low="n"]
````

It doesn't matter what value you give it.
It's rather unfortunate that the README provides no insight as to WTF this setting does.

I'm still puzzled about why this works in 2.4; did someone change how the `configuration.get()` method works? But at any rate, I can put in a fix for this.

**Comment by zeogeo** [comment]
I tried the workaround but for some reason the _active_low_ attribute is ignored from config file, though the validation of config file has no errors.
But I was able to edit the thing by REST API and once the attribute is really there, it works fine.
Any idea why it cannot be added via the config file?

The setting _active_low_ allows to invert the logic of the PIN.  If the _active_low='y'_ then when switched to ON, the PinState is set to LOW (otherwise it would be HIGH).

This settings isn't described anywhere, or at least I didn't see it. Could it be it is also not possible to parse it from the .things config file?
